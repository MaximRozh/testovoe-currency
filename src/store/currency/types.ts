import CurrencyModel from "../../models/Currency.model";

export interface CurrencyState {
  currencies: CurrencyModel[];
  selectedCur: string;
  currenciesCode: string[];
}

export enum CurrencyActionEnum {
  GET_CURRENCY = "GET_CURRENCY",
  SELECT_CURRENCY = "SELECT_CURRENCY",
  SET_CURRENCY_CODE = "SET_CURRENCY_CODE",
}

export interface GetCurrencyAction {
  type: CurrencyActionEnum.GET_CURRENCY;
  payload: CurrencyModel[];
}

export interface SelectCurrencyAction {
  type: CurrencyActionEnum.SELECT_CURRENCY;
  payload: string;
}

export interface SetCurrenciesCodeAction {
  type: CurrencyActionEnum.SET_CURRENCY_CODE;
  payload: string[];
}

export type CurrencyAction =
  | GetCurrencyAction
  | SelectCurrencyAction
  | SetCurrenciesCodeAction;
