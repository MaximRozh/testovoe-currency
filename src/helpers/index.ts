import CurrencyModel from "../models/Currency.model";

export const BASE_CURRENCY = "UAH";
export const BTC_CURRENCY = "BTC";

interface SplitedText {
  amount: number;
  fromCurrency: string;
  toCurrency: string;
}
export const splitedText = (str: string): SplitedText => {
  ///[in]/ig use regEx
  const splited = str.split(" ").filter((item) => item !== "in");
  return {
    amount: +splited[0],
    fromCurrency: splited[1],
    toCurrency: splited[2],
  };
};

export const checkInputText = (
  fromCurrency: string,
  toCurrency: string,
  amount: number
): boolean => {
  const isNotEmpty = fromCurrency && toCurrency && amount;
  const isNotEqual = fromCurrency !== toCurrency;
  const hasCurrencyCode =
    fromCurrency?.length === 3 && toCurrency?.length === 3;
  return amount > 0 && Boolean(isNotEmpty) && isNotEqual && hasCurrencyCode;
};

export const isHaveCurrencyInArray = (
  values: string,
  arr: CurrencyModel[]
): boolean => {
  return arr.some((item) => item.ccy === values?.toUpperCase());
};

export const getCurrencyIndex = (arr: any[], value: string): number =>
  arr.findIndex((item) => item.ccy === value?.toUpperCase());

export const isUah = (val: string): boolean =>
  val?.toUpperCase() === BASE_CURRENCY;

export const isBtc = (val: string): boolean =>
  val?.toUpperCase() === BTC_CURRENCY;

export const toFixed = (val: number): number => Number(val.toFixed(3));

export const validation = (
  fromCurrency: string,
  toCurrency: string,
  amount: number,
  currencies: CurrencyModel[]
) => {
  const isValid = checkInputText(fromCurrency, toCurrency, amount);
  const validCurrencyFrom = isHaveCurrencyInArray(fromCurrency, currencies);
  const validCurrencyTo = isHaveCurrencyInArray(toCurrency, currencies);
  return (
    !isValid ||
    !(
      (validCurrencyFrom || isUah(fromCurrency)) &&
      (validCurrencyTo || isUah(toCurrency))
    )
  );
};
