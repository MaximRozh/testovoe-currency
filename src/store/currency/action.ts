import {
  CurrencyActionEnum,
  GetCurrencyAction,
  SelectCurrencyAction,
  SetCurrenciesCodeAction,
} from "./types";

import CurrencyModel from "../../models/Currency.model";
import { AppDispatch } from "..";
import CurrencyService from "../../services/currencyService";

export const CurrencyActionCreator = {
  setCurencies: (data: CurrencyModel[]): GetCurrencyAction => ({
    type: CurrencyActionEnum.GET_CURRENCY,
    payload: data,
  }),
  setCurrenciesCode: (data: string[]): SetCurrenciesCodeAction => ({
    type: CurrencyActionEnum.SET_CURRENCY_CODE,
    payload: data,
  }),
  //try to fix ANY
  getCurencies:
    (): any =>
    (dispatch: AppDispatch): void => {
      CurrencyService.getCurrency().then((res) => {
        dispatch(CurrencyActionCreator.setCurencies(res));

        const currenciesCode = res.map((item: CurrencyModel) => item.ccy);
        dispatch(
          CurrencyActionCreator.setCurrenciesCode(["UAH", ...currenciesCode])
        );
      });
    },
  selectCurrency: (currency: string): SelectCurrencyAction => ({
    type: CurrencyActionEnum.SELECT_CURRENCY,
    payload: currency,
  }),
};
