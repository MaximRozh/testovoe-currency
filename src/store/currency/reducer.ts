import { CurrencyState, CurrencyActionEnum, CurrencyAction } from "./types";

const initialState: CurrencyState = {
  currencies: [],
  selectedCur: "UAH",
  currenciesCode: [],
};

export default function currencyReducer(
  state = initialState,
  action: CurrencyAction
): CurrencyState {
  switch (action.type) {
    case CurrencyActionEnum.GET_CURRENCY:
      return { ...state, currencies: action.payload };
    case CurrencyActionEnum.SELECT_CURRENCY:
      return { ...state, selectedCur: action.payload };
    case CurrencyActionEnum.SET_CURRENCY_CODE:
      return { ...state, currenciesCode: action.payload };
    default:
      return state;
  }
}
