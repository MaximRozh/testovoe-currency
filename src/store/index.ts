import { applyMiddleware, combineReducers, createStore } from "redux";
import thunk from "redux-thunk";
import currencyReducer from "./currency/reducer";

const rootReducer = combineReducers({
  currencyReducer,
});

export const store = createStore(rootReducer, applyMiddleware(thunk));

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
