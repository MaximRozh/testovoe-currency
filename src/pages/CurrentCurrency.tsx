import React, { FC, useEffect, useMemo } from "react";
import { useDispatch } from "react-redux";
import ContentHeader from "../components/ContentHeader";
import Layout from "../components/Layout";
import { BASE_CURRENCY, isBtc, toFixed } from "../helpers";
import { useTypedSelector } from "../hooks/useTypedSelector";
import { CurrencyActionCreator } from "../store/currency/action";

const CurrentCurrency: FC = () => {
  const { currencies, selectedCur } = useTypedSelector(
    (state) => state.currencyReducer
  );

  const dispatch = useDispatch();
  useEffect(() => {
    if (!currencies.length) {
      dispatch(CurrencyActionCreator.getCurencies());
    }
  }, [dispatch, currencies]);

  const ratesToCyrrency = useMemo(() => {
    const rate = currencies.find((item) => item.ccy === selectedCur)?.buy || 1;

    return currencies.map((item) => {
      let obj = { ...item };
      const isItemSelectedCyrrency = item.ccy === selectedCur;

      if (isBtc(item.ccy)) {
        const usdRate = currencies.find((item) => item.ccy === "USD")?.buy || 1;
        obj = { ...item, buy: toFixed((item.buy * usdRate) / rate) };
      }
      if (isItemSelectedCyrrency) {
        obj = { ccy: BASE_CURRENCY, buy: toFixed(1 / rate) };
      }
      if (!isItemSelectedCyrrency && !isBtc(item.ccy)) {
        obj = { ...item, buy: toFixed(item.buy / rate) };
      }

      return obj;
    });
  }, [currencies, selectedCur]);

  return (
    <Layout>
      <div className="currency-wrapper">
        <div className="foo">
          <div className="header-wrapper">
            {" "}
            <ContentHeader content="exchange rates" />
          </div>
          <div className="currency-rates">
            <div className="currency-rates-header">
              <span className="rates-header-text">Currency</span>
              <span className="rates-header-text">Amount</span>
            </div>
            <ul className="all-rates">
              {ratesToCyrrency.map((item) => (
                <li className="rate-item" key={item.ccy}>
                  <span className="rate-text">{item.ccy} :</span>{" "}
                  <span className="rate-text">{item.buy}</span>
                </li>
              ))}
            </ul>
          </div>
        </div>
      </div>
    </Layout>
  );
};

export default CurrentCurrency;
