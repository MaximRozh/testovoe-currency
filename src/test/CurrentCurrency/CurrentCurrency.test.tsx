import React from "react";
import { cleanup, RenderResult } from "@testing-library/react";
import { store } from "../../store";
import { CurrencyActionCreator } from "../../store/currency/action";
import { act } from "react-dom/test-utils";
import { toFixed } from "../../helpers";
import {
  currencies,
  ExchangeCurrency,
  getRatesFromCurrencies,
} from "../testUtils";
import { renderWithProviders } from "../renderWithProvider";
import CurrentCurrency from "../../pages/CurrentCurrency";

jest.useFakeTimers("modern");
let div: RenderResult;

describe("Check Current Rates", () => {
  beforeEach(() => {
    div = renderWithProviders(<CurrentCurrency />);
    act(() => {
      store.dispatch(CurrencyActionCreator.setCurencies(currencies));
    });
  });
  afterEach(cleanup);

  it("All rates are displayed", () => {
    const retes = div.container.querySelectorAll(".rate-item");
    expect(retes.length).toBe(currencies.length);
  });
  it("If select USD - change rates to UAH", () => {
    act(() => {
      store.dispatch(CurrencyActionCreator.selectCurrency("USD"));
    });
    const retes = div.container.querySelectorAll(".rate-item");

    const { USD } = getRatesFromCurrencies([ExchangeCurrency.USD], currencies);

    const changedCurrencyCode = retes[0].firstElementChild?.innerHTML.replace(
      " :",
      ""
    );
    const changedRate = retes[0].lastChild?.textContent;

    expect(changedCurrencyCode).toBe("UAH");
    expect(changedRate).toBe(toFixed(1 / USD).toString());
  });
});

describe("CurrentCurrency snapshot", () => {
  it("Make CurrentCurrency snapshot", () => {
    const div = renderWithProviders(<CurrentCurrency />);
    const { asFragment } = div;
    expect(asFragment()).toMatchSnapshot();
  });
});
