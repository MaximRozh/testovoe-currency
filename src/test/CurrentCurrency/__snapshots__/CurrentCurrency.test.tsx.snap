// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Current Rates Make ConvertCurrency snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="nav-wrapper"
    >
      <div
        class="nav-wrapper-container"
      >
        <nav
          class="nav-bar"
        >
          <a
            aria-current="page"
            class="active"
            href="#/"
          >
            <img
              alt="Logo"
              class="logo"
              src="Label.svg"
            />
          </a>
          <ul
            class="nav-menu"
          >
            <li
              class="nav-menu-item"
            >
              <a
                aria-current="page"
                class="active-link"
                href="#/"
              >
                Currency converter
              </a>
            </li>
            <li
              class="nav-menu-item"
            >
              <a
                class=""
                href="#/current-currency"
              >
                Exchange rates
              </a>
            </li>
          </ul>
          <select />
        </nav>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="currency-wrapper"
      >
        <div
          class="foo"
        >
          <div
            class="header-wrapper"
          >
             
            <h2
              class="convert-header"
            >
              EXCHANGE RATES
            </h2>
            <p
              class="convert-text"
            >
              Fast, simplle, non-custodial Exchange Service. Check live foreign currency exchange rates
            </p>
          </div>
          <div
            class="currency-rates"
          >
            <div
              class="currency-rates-header"
            >
              <span
                class="rates-header-text"
              >
                Currency
              </span>
              <span
                class="rates-header-text"
              >
                Amount
              </span>
            </div>
            <ul
              class="all-rates"
            >
              <li
                class="rate-item"
              >
                <span
                  class="rate-text"
                >
                  USD :
                </span>
                 
                <span
                  class="rate-text"
                >
                  29.255
                </span>
              </li>
              <li
                class="rate-item"
              >
                <span
                  class="rate-text"
                >
                  EUR :
                </span>
                 
                <span
                  class="rate-text"
                >
                  30.255
                </span>
              </li>
              <li
                class="rate-item"
              >
                <span
                  class="rate-text"
                >
                  RUR :
                </span>
                 
                <span
                  class="rate-text"
                >
                  0.32
                </span>
              </li>
              <li
                class="rate-item"
              >
                <span
                  class="rate-text"
                >
                  BTC :
                </span>
                 
                <span
                  class="rate-text"
                >
                  883832.156
                </span>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
